python/cudf/cudf/io/parquet.py

```diff
@@ -3,12 +3,106 @@
 import warnings
 
 import pyarrow.parquet as pq
+from pyarrow.compat import guid
 
 import cudf
 import cudf._libxx.parquet as libparquet
 from cudf.utils import ioutils
 
 
+def _get_partition_groups(df, partition_cols, preserve_index=False):
+    df = df.sort_values(partition_cols)
+    if not preserve_index:
+        df = df.reset_index(drop=True)
+    divisions = df[partition_cols].drop_duplicates()
+    splits = df[partition_cols].searchsorted(divisions, side="left")
+    splits = splits.tolist() + [len(df[partition_cols])]
+    return [
+        df.iloc[splits[i] : splits[i + 1]].copy(deep=False)
+        for i in range(0, len(splits) - 1)
+    ]
```

Comment:
Ideally this logic should move down into the C++ in the future.

Reply:
cc @OlivierNV this is logic for partitioning the dataframe based on a set of columns, but I imagine we could implement it much more efficiently in libcudf, especially if we need to output a large number of partitions

Reply:
Is this for cropping the dataframe by rows (slicing), or to set specific rowgroup boundaries ?

Reply:
Looking at the python code, that seems like some fancy partitioning that could be done more efficiently in fewer passes with a dedicated kernel, though it probably would have to still be upstream from the writer (writer needs contiguous column data to be efficient).

Reply:
> Is this for cropping the dataframe by rows (slicing), or to set specific rowgroup boundaries ?

Even in python, this code is likely more complex than it needs to be.  With that said, I personally ran into issues a while back when I tried to simplify it.

The idea is to effectively do a local `groupby` within each dataframe partition, and then write each "group" to a separate file.  In pyarrow, this is literally done with a pandas groupby operation.  However, we cannot loop through groups in the same way with cudf (or at least we couldn't when I wrote this code), so I just used the sort and split approach.

Reply:
This will change with #4346 but it may be a bit before it's ready to merge. Regardless, I have concerns of calling `to_parquet` multiple times, and it would be more ideal if we could somehow pass a single sorted table, the indices into the groups, and a vector of output locations to the writer all at once so that there's additional opportunities to parallelize.

Reply:
> This will change with #4346 but it may be a bit before it's ready to merge. Regardless, I have concerns of calling to_parquet multiple times, and it would be more ideal if we could somehow pass a single sorted table, the indices into the groups, and a vector of output locations to the writer all at once so that there's additional opportunities to parallelize.

Makes sense to me.  Is the suggestion then to add a `"TODO"` here and open an issue for lower-level partitioned-write support?